### ISSUE TITLE
GQ signature support

### ISSUE THREAD

**Issue by jonnystoten** [bastionzero/openpubkey#7]
Currently, an OpenPubkey token contains the original signature from the OIDC provider. This arguably means that it isn't safe to share the OpenPubkey token, as the token could be used to impersonate the subject to a service that doesn't check the audience claim, or to trick an OpenPubkey client to issue a token for the subject.

We can create a GQ signature to prove that we have the original OIDC provider's signature. This can be verified using the original OIDC signing payload (`header || '.' || payload`) and the OIDC provider's public key. This works because GQ signature private keys are equivalent to RSA signatures.

The GQ signature can be published in place of the OIDC provider's signature in an OpenPubkey token.

**Comment by anderspitman** [comment]
@jonnystoten considering using OpenPubKey in a project and want to make sure I understand the security implications here. Providing the original ID token to potentially untrusted parties is only dangerous if there exists a vulnerability in some other system, right?

For example, if there's an app out there where the frontend does a Google OIDC flow, then hands the ID token to the backend in order to create a session, a leaked OpenPubKey ID token could be used to log in to that app?

So assuming apps properly check `aud`, this shouldn't be an issue?

**Comment by kipz** [comment]
@anderspitman that's right, and that wouldn't be without precedent (e.g. https://github.com/dgrijalva/jwt-go/issues/422). Even if they're just bugs in the app though, this protects against them.

**Comment by anderspitman** [comment]
Awesome, thanks!

**Comment by jonnystoten** [comment]
Yep, what @kipz said. For completeness, also note that this issue is resolved with the support for GQ signatures, see

- #3
- #14

**Comment by anderspitman** [comment]
Yeah I hadn't heard of GQ signatures but they look very interesting. Unfortunately I don't think OpenPubKey will work for my use case. I was hoping to build a generic authn system where you could sign in to an app once with any OIDC provider, using OPK, then send that ID token (maybe with a GQ sig) along with requests to other servers, with the requests signed with DPoP. The other servers know that you have a key pair certified to be controlled by the OIDC user.

The problem is that you're essentially giving such an app permissions to be you on any backend that supports the protocol. With a generic OIDC provider, there's not really any way to explain to users what they're consenting too. I don't think many devs would or should build backends to respect such a system.

Basically it boils down to I think I need to design an OIDC profile special built for this purpose. Something along the lines of Solid-OIDC I think.

**Comment by EthanHeilman** [comment]
@anderspitman 

> The problem is that you're essentially giving such an app permissions to be you on any backend that supports the protocol. 

Could you use the client ID to scope this? Or have the local app ask the user for consent?

> Basically it boils down to I think I need to design an OIDC profile special built for this purpose. Something along the lines of Solid-OIDC I think.

Did you make any progress on this, it sounds useful

**Comment by anderspitman** [comment]
> Could you use the client ID to scope this?

You mean configuring services to only accept requests with tokens originating from approved clients? That would work for major apps, but might severely limit self-hosted clients. Every backend instance would have to approve every client. Maybe not the end of the world though, and would probably still be a nice step in the right direction.

> Or have the local app ask the user for consent?

Not sure this would work. Malicious apps are the problem.

My goal is for me to host serverA.com, and my wife to host serverB.com. I want to be able to log in to app1.com with OpenPubKey, and send requests to serverA.com and serverB.com with no prior interaction with those servers, other than to tell them my OIDC identifier is authorized to make requests.

The problem is if I log in to evil.com with the same OIDC provider, it could mint OPK tokens and make requests to serverA.com and serverB.com without me ever knowing about it.

> Did you make any progress on this, it sounds useful

It's still in my backlog sadly, but if others are interested in collaborating I would certainly be interested in working on it sooner.

**Comment by EthanHeilman** [comment]
> Not sure this would work. Malicious apps are the problem.

Ok, so the idea is that any app can talk to any backend and it should just work? The problem is OIDC is built on the assumption that serverA.com and serverB.com each have an allow list of approved apps in the form a approved client-IDs.

If OIDC has a built in mechanism to say "STOP, do you want to give javascript evil.example.com the ability to sign messages on your behalf to any other server or website."  Would that solve your problem?

I think there are two ways to build this with OpenPubkey but they aren't off-the-shelf:

1. Use the MFA-cosigner and add such a message to the MFA-Cosigner. This wouldn't also remove the trust assumption on the IDP. The OpenPubkey opensource project has a prototype cosigner, but it isn't production ready.

2. Have a single approved client-id that is just part of setting OpenPubkey up on a server. Then have javascript client at an origin which is the origin that the IDP will send ID Token to. Any webapp that wants to use OpenPubkey would need to iframe the javascript client and the javascript client would intermediate any signing request. Thus, the javascript client could require a confirmation from the user and/or restrict apps using OpenPubkey.

Both 1 and 2 are buildable today. 2 would be much more work because all the OpenPubkey code is golang.

**Comment by anderspitman** [comment]
> If OIDC has a built in mechanism to say "STOP, do you want to give javascript evil.example.com the ability to sign messages on your behalf to any other server or website." Would that solve your problem?

Yeah I think that's pretty much what we're looking for

1. I need to read up on the MFA-cosigner. I've heard it mentioned before but don't know much about how it works.

2. If I'm understanding this correctly, it would be fairly centralized, right? ie whoever controls that shared client could inject malicious JS, go out of business, be really slow etc at any point? I suppose you could have multiple instances run by different orgs though. 
This sort of reminds me of Mozilla Persona, which was partially centralized in practice, but I think ideally needs to be built into browsers.
 
3.I wonder if the UX would be confusing since the IdP would say you're logging in to the special JS shim, not the actual app. My project [LastLogin](https://lastlogin.net/) has this problem. You can use it to add Google login to your app/site without setting up an OIDC client (or even a Google account), but when the user signs in Google tells them they're signing in with LastLogin. It's not the end of the world because you can show a "Sign in with Google via LastLogin" to your site, but it's still more confusing than a traditional setup.

> I'd be interested, for sure!

I'll shoot you an email!

**Comment by EthanHeilman** [comment]
> If I'm understanding this correctly, it would be fairly centralized, right? ie whoever controls that shared client could inject malicious JS, go out of business, be really slow etc at any point? I suppose you could have multiple instances run by different orgs though.

In chrome there are some tricks that let to do have a 3 JS apps that each refuse to read from their secret store if the other two have code which doesn't match the code they expect. If you did some clever MPC on the OIDC verification code, you could ensure security  and availability against 1 of the 3 apps being malicious. That approach is complex and would be its own project.

> I wonder if the UX would be confusing since the IdP would say you're logging in to the special JS shim, not the actual app.

I hadn't thought of that.

**Comment by anderspitman** [comment]
> In chrome there are some tricks that let to do have a 3 JS apps that each refuse to read from their secret store if the other two have code which doesn't match the code they expect. If you did some clever MPC on the OIDC verification code, you could ensure security and availability against 1 of the 3 apps being malicious. That approach is complex and would be its own project.

That's really interesting. Do you have any links or search terms where I can read more about that?

**Comment by EthanHeilman** [comment]
It isn't well documented anywhere, but was an observation I made. 

You need [CSP embedded enforcement (cspee)](https://w3c.github.io/webappsec-cspee/), which is supported in chromium based browsers.

You have 3 iframes, each running at a different origin. They each use cspee to ensure that the other two are running the code they expect. Basically SRI but they don't need to load each other.  They can also get SRI like guarantees on the page which loads the iframes.

They can pass messages to each other.  This allows them to use secret sharing to split secret keys between them and than do MPC over the secrets to jointly sign statements they at least agree with.

You split the user's signing key into three parts,, send one to each of the iframes and 2 of the 3 must agree to generate a signature. This should be [very fast because it is just ECC math.](https://eprint.iacr.org/2020/1261)

I tried to get firefox to support cspee a while ago, but I didn't have the time to write a patch
https://github.com/mozilla/standards-positions/issues/326
